Frameworks/Foundation/NSURLConnection.mm

```diff
@@ -346,118 +373,152 @@ - (void)cancel {
+    __dispatchDelegateCallback(self, ^void() {
+        if ([strongSelf->_delegate respondsToSelector:@selector(connection:didReceiveData:)]) {
+            [strongSelf->_delegate connection:strongSelf didReceiveData:strongData];
+        }
+    });
 }
 
 /**
  @Status Interoperable
 */
-- (void)URLProtocolDidFinishLoading:(id)urlProtocol {
-    /*
-    if(_storagePolicy==NSURLCacheStorageNotAllowed) {
-    //[[NSURLCache sharedURLCache] removeCachedResponseForRequest:_request];
-    } else {
-    //NSCachedURLResponse *cachedResponse=[[NSCachedURLResponse alloc] initWithResponse:_response data:_mutableData
-    userInfo:nil storagePolicy:_storagePolicy];
-
-    //if([_delegate respondsToSelector:@selector(connection:willCacheResponse:)])
-    //cachedResponse=[_delegate connection:self willCacheResponse:cachedResponse];
-
-    //if(cachedResponse!=nil){
-    //[[NSURLCache sharedURLCache] storeCachedResponse:cachedResponse forRequest:_request];
-    //}
-    }
-    */
+- (void)URLProtocolDidFinishLoading:(NSURLProtocol*)urlProtocol {
+    __block StrongId<NSURLConnection> strongSelf = self;
+    __dispatchDelegateCallback(self, ^void() {
+        /*
+        if(_storagePolicy==NSURLCacheStorageNotAllowed) {
+        //[[NSURLCache sharedURLCache] removeCachedResponseForRequest:_request];
+        } else {
+        //NSCachedURLResponse *cachedResponse=[[NSCachedURLResponse alloc] initWithResponse:_response data:_mutableData
+        userInfo:nil storagePolicy:_storagePolicy];
 
-    if ([_delegate respondsToSelector:@selector(connectionDidFinishLoading:)]) {
-        [_delegate performSelector:@selector(connectionDidFinishLoading:) withObject:self];
-    }
+        //if([_delegate respondsToSelector:@selector(connection:willCacheResponse:)])
+        //cachedResponse=[_delegate connection:self willCacheResponse:cachedResponse];
+
+        //if(cachedResponse!=nil){
+        //[[NSURLCache sharedURLCache] storeCachedResponse:cachedResponse forRequest:_request];
+        //}
+        }
+        */
 
-    _protocol = nil;
-    _delegate = nil;
+        if ([strongSelf->_delegate respondsToSelector:@selector(connectionDidFinishLoading:)]) {
+            [strongSelf->_delegate connectionDidFinishLoading:strongSelf];
+        }
+
+        strongSelf->_protocol = nil;
+        strongSelf->_delegate = nil;
+        strongSelf->_done = true;
+    });
 }
 
 /**
  @Status Interoperable
 */
-- (void)URLProtocol:(id)urlProtocol didReceiveAuthenticationChallenge:(id)challenge {
-    if ([_delegate respondsToSelector:@selector(connection:willSendRequestForAuthenticationChallenge:)]) {
-        [_delegate connection:self willSendRequestForAuthenticationChallenge:challenge];
-    } else {
-        [_delegate connection:self didReceiveAuthenticationChallenge:challenge];
-    }
+- (void)URLProtocol:(NSURLProtocol*)urlProtocol didReceiveAuthenticationChallenge:(NSURLAuthenticationChallenge*)challenge {
+    __block StrongId<NSURLConnection> strongSelf = self;
```

Comment:
I believe using __block on these _avoids_ copying the `StrongId` -- self is therefore _not_ strongly held by the block.